ChronoAPI/ChronoLog/include/rpc.h

```diff
@@ -146,40 +146,13 @@ class ChronoLogRPC {
                 case CHRONOLOG_THALLIUM_TCP:
                 case CHRONOLOG_THALLIUM_SOCKETS:
                 case CHRONOLOG_THALLIUM_ROCE: {
-                    hg_addr_t addr_self;
-                    hg_return_t hret;
-                    for (size_t i = 0; i < workers; i++) {
-                        LOGD("creating Thallium server with engine str %s", serverAddrList_[i].c_str());
-                        margo_instance_id mid = margo_init(serverAddrList_[i].c_str(), MARGO_SERVER_MODE,
-                                                           1, numStreams_);
-                        if (mid == MARGO_INSTANCE_NULL) {
-                            LOGE("Error: margo_init()");
-                            exit(-1);
-                        }
-
-                        /* figure out first listening addr */
-                        hret = margo_addr_self(mid, &addr_self);
-                        if (hret != HG_SUCCESS) {
-                            LOGE("Error: margo_addr_self()");
-                            margo_finalize(mid);
-                            exit(-1);
-                        }
-                        hg_size_t addr_self_string_sz = 128;
-                        char addr_self_string[128];
-                        hret = margo_addr_to_string(mid, addr_self_string, &addr_self_string_sz,
-                                                    addr_self);
-                        if (hret != HG_SUCCESS) {
-                            LOGE("Error: margo_addr_to_string()");
-                            margo_addr_free(mid, addr_self);
-                            margo_finalize(mid);
-                            exit(-1);
-                        }
-                        margo_addr_free(mid, addr_self);
-
-                        tl::engine new_engine(mid);
-                        thalliumServerList_.emplace_back(std::move(new_engine));
-                        LOGI("engine: %s is created", std::string(thalliumServerList_[i].self()).c_str());
+                    for (size_t i = 0; i < workers; i++) 
+		    {
+			tl::engine *tmpserver = new tl::engine(serverAddrList_[i].c_str(),THALLIUM_SERVER_MODE,true,2);
+			thalliumServerList_.push_back(tmpserver);
```

Comment:
You remove the multi-port engine code. If we do not want to have Visor to listen on multiple ports. Do we need to have multiple engines at all?

Reply:
@fkengun We should have multiple engines. But, this looked cleaner and it seemed to work. Maybe, the new thallium version fixed the bug we had previously. We should probably test it on Ares and see if it works with hundreds of thallium engines. It did not break on my laptop. There is an issue with using smart pointers with ChronoLogRPC class which is different from Hari's usage. In our case, the server doesn't stop. Its destructor doesn't run to completion and that was causing the segmentation faults. This may be a workaround, until we find a better solution.

Reply:
@aparnasasidharan2017 the code does look cleaner , but what was the bug you both are referring to? was it smth about  the shared_ptr,  the vector of engines vs the vector of engine pointers, or the failure of individual engine to finalize properly ?

Reply:
@aparnasasidharan2017 I don't think Thallium version matters. I've been using the latest Thallium@0.10.1 for a while. It did not change anything.
@ibrodkin The bug we are referring to is that using class-local member variables to store ClientRegistryManager and ChronicleMetaDirectory segs fault. All the variables get wiped between initialization and handling first request. That is why I used the ugly global variables. Now @aparnasasidharan2017 finds a way to do it properly using member variables in RPCVisor.
@aparnasasidharan2017 I think you switched from the Singleton-based creation to traditional new for the creation of ClientRegistryManager and ChronicleMetaDirectory. That is more likely to be the key to me.

Reply:
@fkengun I have made some additional changes and bug fixes. Also, updated my margo version to 0.11.1 and thallium to 0.10.1. There was a bug in the Visor APIs which was causing a deadlock in the Visor when the number of service threads > 1. I have removed the deadlock. The code works with multiple RPC service threads, but there was a margo exception once. I'll look into that tomorrow. I have to work on the aesthetics and code alignment also.

Reply:
The margo exceptions are rare, they occur when you issue a large number of  RPC requests too soon. It is not due to my code changes. The exceptions occur with the previous engine initialization method also. We should put that under known issues.

Reply:
I have made some changes to the CMakefiles to support intel-gcc and intel-llvm compilation chains. Our current software stack works with intel compilers also. The client test cases did not fail with icc (classic) or with intel HPC icx compilers. You can enable them by uncommenting two lines in CMakelists.txt.